org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/editor/quickfix/AbstractIdeQuickfixTest.xtend

```diff
@@ -0,0 +1,174 @@
+		val elementNode = NodeModelUtils.getNode(target)
+
+		val elementStartPosition = NodeModelUtils.getLineAndColumn(elementNode, elementNode.offset)
+		val elementEndPosition = NodeModelUtils.getLineAndColumn(elementNode, elementNode.endOffset)
+
+		val startPos = new Position(elementStartPosition.line - 1, elementStartPosition.column - 1)
+		val endPos = new Position(elementEndPosition.line - 1, elementEndPosition.column - 1)
+
+		val issue = new Diagnostic => [
+			code = issueCode
+			message = "error"
+			severity = DiagnosticSeverity.Error
+			source = "source"
+			range = new Range(startPos, endPos)
+		]
+
+		val options = new ICodeActionService2.Options()
+		options.cancelIndicator = CancelIndicator.NullImpl
+		options.document = new Document(0, toUnixLineSeparator(originalText))
+		options.resource = target.eResource as XtextResource
+
+		options.languageServerAccess = new ILanguageServerAccess() {
+
+			override addBuildListener(IBuildListener listener) {
+				// stub
+			}
+
+			override <T> doRead(String uri, Function<Context, T> function) {
+				val ctx = new ILanguageServerAccess.Context(options.resource, options.document,
+					true, CancelIndicator.NullImpl)
+				CompletableFuture.completedFuture(function.apply(ctx));
+			}
+
+			override <T> doReadIndex(Function<? super IndexContext, ? extends T> function) {
+				null
+			}
+
+			override getInitializeParams() {
+				null
+			}
+
+			override getInitializeResult() {
+				null
+			}
+
+			override getLanguageClient() {
+				null
+			}
+
+			override newLiveScopeResourceSet(URI uri) {
+				options.resource.resourceSet
+			}
+
+		}
+
+		options.codeActionParams = new CodeActionParams => [
+			context = new CodeActionContext => [
+				diagnostics = #[issue]
+			]
+			range = new Range(startPos, endPos)
+			textDocument = new TextDocumentIdentifier(target.eResource.URI.toString)
+		]
+
+		val actualIssueResolutions = quickFixProvider.getResolutions(options, issue).sortBy[label]
+		assertEquals("The number of quickfixes does not match!", expectedSorted.size,
+			actualIssueResolutions.size)
+		for (i : 0 ..< actualIssueResolutions.size) {
+			val actualIssueResolution = actualIssueResolutions.get(i)
+			val expectedIssueResolution = expectedSorted.get(i)
+			assertEquals(expectedIssueResolution.label, actualIssueResolution.label)
+			// TODO description?
+			assertEquals(expectedIssueResolution.description, actualIssueResolution.label)
+			assertIssueResolutionResult(toUnixLineSeparator(expectedIssueResolution.result), actualIssueResolution,
+				originalText, options.document)
+		}
+	}
+
+	protected def void assertIssueResolutionResult(String expectedResult,
+		DiagnosticResolution actualIssueResolution, String originalText, Document doc) {
+		val edit = actualIssueResolution.apply
+		val changedDocument = doc.applyChanges(edit.changes.values.flatMap[toList])
+
+		assertEquals(expectedResult, changedDocument.contents)
+	}
+
+	@Data
+	protected static class Quickfix {
+
```

Comment:
Nitpick probably don't need any empty lines in this `@Data`, it's supposed to be very concise